libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/OnlineRouteAlternativesSwitch.kt

```diff
@@ -0,0 +1,268 @@
+        mapboxNavigationScope.launch {
+            mapboxNavigation.flowRoutesUpdated()
+                .map { it.navigationRoutes }
+                .requestOnlineRoutes(
+                    mapboxNavigation.flowLocationMatcherResult(),
+                    mapboxNavigation.flowRouteProgress(),
+                    routeRequestMechanism = { options ->
+                        requestRoutes(
+                            options,
+                            accessToken,
+                            MapboxNavigationAccounts,
+                            connectTimeoutMilliseconds = connectTimeoutMilliseconds,
+                            readTimeoutMilliseconds = readTimeoutMilliseconds
+                        )
+                    },
+                    minimumRetryInterval = minimumRetryInterval.toLong()
+                )
+                .collectLatest {
+                    mapboxNavigation.setNavigationRoutes(it)
+                }
+        }
+    }
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigationScope.cancel()
+    }
+}
+
+internal sealed class DirectionsRequestResult {
+    data class SuccessfulResponse(val body: DirectionsResponse) : DirectionsRequestResult()
+    object Error : DirectionsRequestResult()
+}
+
+internal typealias RouteRequestMechanism = suspend (RouteOptions) -> DirectionsRequestResult
+
+@OptIn(ExperimentalCoroutinesApi::class)
+internal fun Flow<List<NavigationRoute>>.requestOnlineRoutes(
+    locations: Flow<LocationMatcherResult>,
+    routeProgress: Flow<RouteProgress>,
+    routeRequestMechanism: RouteRequestMechanism,
+    minimumRetryInterval: Long = 60_000,
+    navigationRouteSerializationDispatcher: CoroutineDispatcher = Dispatchers.Default
+): Flow<List<NavigationRoute>> = mapLatest {
+    if (it.isEmpty()) {
+        null
+    } else {
+        val primaryRoute = it.first()
+        if (primaryRoute.origin == Onboard) {
+            logI(LOG_CATEGORY) { "Current route is offline, requesting online routes" }
+            requestOnlineRouteWithRetryOrNull(
+                locations,
+                primaryRoute,
+                routeProgress,
+                routeRequestMechanism,
+                navigationRouteSerializationDispatcher,
+                minimumRetryInterval
+            )
+        } else {
+            null
+        }
+    }
+}.filterNotNull()
+
+private suspend fun requestOnlineRouteWithRetryOrNull(
+    locations: Flow<LocationMatcherResult>,
+    primaryRoute: NavigationRoute,
+    routeProgress: Flow<RouteProgress>,
+    routeRequestMechanism: RouteRequestMechanism,
+    navigationRouteSerializationDispatcher: CoroutineDispatcher,
+    minimumRetryInterval: Long
+): List<NavigationRoute>? = coroutineScope {
+    while (true) {
+        val retryLimiter = launch { delay(minimumRetryInterval) }
+        try {
+            val latestPosition = locations.first()
+            val latestRouteProgress = routeProgress.first()
```

Comment:
Can we add a test that would check that the latest location is used in case of retries?